tests/tools/test_validate_tool_call.py

```diff
@@ -0,0 +1,184 @@
+            required_tools=["tool1", "tool2"],
+            actual_tool_calls=[],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Not all required tools were called")
+
+    def test_all_required_tools_called_in_order(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2"],
+            actual_tool_calls=["tool1", "tool2"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_required_tools_called_out_of_order(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2"],
+            actual_tool_calls=["tool2", "tool1"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertFalse(result)
+
+    def test_forbidden_tool_called(self):
+        result, message = validate_tool(
+            required_tools=["tool1"],
+            actual_tool_calls=["tool1", "tool3"],
+            optional_tools=[],
+            forbidden_tools=["tool3"],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Forbidden tool called: tool3")
+
+    def test_optional_tools_called(self):
+        result, message = validate_tool(
+            required_tools=["tool1"],
+            actual_tool_calls=["tool1", "tool2"],
+            optional_tools=["tool2"],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_unexpected_tool_called(self):
+        result, message = validate_tool(
+            required_tools=["tool1"],
+            actual_tool_calls=["tool1", "tool3"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Unexpected tool called: tool3")
+
+    def test_all_required_tools_called_with_optional_and_forbidden(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2"],
+            actual_tool_calls=["tool1", "tool2", "tool3"],
+            optional_tools=["tool3"],
+            forbidden_tools=["tool4"],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_no_required_tools(self):
+        result, message = validate_tool(
+            required_tools=[],
+            actual_tool_calls=["tool1", "tool2"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertFalse(result)
+
+    def test_no_required_tools_but_optional(self):
+        result, message = validate_tool(
+            required_tools=[],
+            actual_tool_calls=["tool1", "tool2"],
+            optional_tools=["tool1", "tool2"],
```

Comment:
Does/should the order of the optional tools matter ? Real case exemple:
get features from a certain morpho with tortuosity < 1:
- `required: [resolve]`
- `optional: [get_morpho, morpho_feature, kg_morpho_feature]`
Here `get_morpho` should happen before `morpho_feature`, but `kg_morpho_feature` can happen anytime in theory.

Reply:
so either it can fetch morph and calculate by itself or it can fetch from kg directly. so maybe both options should be ok

Reply:
Lets keep this in mind in future